### ISSUE TITLE
AMAZON_AWS_S3_PKG "Duplicate Headers"

### ISSUE THREAD

**Issue by jeffreykemp** [mortenbra/alexandria-plsql-utils#83]
Recently my code started failing with an error "Duplicate Headers" being returned from AWS, e.g. when uploading an object using AMAZON_AWS_S3_PKG.

Note: I was able to resolve this by editing the package to remove the bits that set the "Host" header. I don't know why that fixed it but it seems to work fine without this.

```  --l_header_names.extend;
  --l_header_names(1) := 'Host';
  --l_header_values.extend;
  --l_header_values(1) := g_aws_host_s3;`

(this edit required adjusting the array indices for the other headers)

**Comment by Jeremiah-England** [comment]
Just linking another discussion of the same issue: https://repost.aws/questions/QUUyrvzv0GRSyRQb7T9gXZhQ/error-duplicate-header-name-upon-put-to-s-3-bucket.

**Comment by mortenbra** [comment]
Thanks for the feedback!

I also started noticing this some weeks ago (actually, I can pinpoint the problem quite exactly to start appearing on February 24, 2023). I've had the same code running every day for many years doing backups, on the same "client" (database), so obviously this was something that changed on the AWS side. I looked at their docs and blogs, but could not find any mentions.

I would variously see the error "Your request contains duplicate headers." returned from AWS, at other times I would just get "ORA-29273: HTTP request failed" lower down the stack. From what I could observe, uploading very small files would still work, while larger files fail.

If the permanent fix is indeed to remove the "Host" header, I think I'll do a small refactoring of how the headers are set and commit a new version to Git when I have some free time.

**Comment by mrtrollex** [comment]
thanks!
same error for me,i removed host header and reindexed rest of headers.

But now im getting - SignatureDoesNotMatch error

Does somebody have fix for that?
Thanks

**Comment by mrtrollex** [comment]
i found out that changing 
                                           `'HTTP/1.1'); `
in
```
utl_http.begin_request(get_url(p_bucket_name, l_key),
                                           'PUT',
                                           'HTTP/1.1');
```

to
 ```
utl_http.begin_request(get_url(p_bucket_name, l_key),
                                           'PUT',
                                           'HTTP/1.0');
```
also fixes duplicate header error, because 1.1 is automatically setting host header
but it still doesnt fix my SignatureDoesNotMatch error

**Comment by mortenbra** [comment]
> thanks! same error for me,i removed host header and reindexed rest of headers.
> 
> But now im getting - SignatureDoesNotMatch error
> 
> Does somebody have fix for that? Thanks

Are you calling amazon_aws_auth_pkg.init() first?

**Comment by mrtrollex** [comment]
> > thanks! same error for me,i removed host header and reindexed rest of headers.
> > But now im getting - SignatureDoesNotMatch error
> > Does somebody have fix for that? Thanks
> 
> Are you calling amazon_aws_auth_pkg.init() first?

Yes and my credentials are correct, i think its about not correct headers in signature? Not so sure..
